packages/edit-site/src/hooks/template-part-edit.js

```diff
@@ -0,0 +1,82 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useSelect } from '@wordpress/data';
+import { BlockControls } from '@wordpress/block-editor';
+import { store as coreStore } from '@wordpress/core-data';
+import { ToolbarButton } from '@wordpress/components';
+import { addFilter } from '@wordpress/hooks';
+import { createHigherOrderComponent } from '@wordpress/compose';
+
+/**
+ * Internal dependencies
+ */
+import { useLocation } from '../components/routes';
+import { useLink } from '../components/routes/link';
+
+function EditTemplatePartMenuItem( { attributes } ) {
+	const { theme, slug } = attributes;
+	const { params } = useLocation();
+	const templatePart = useSelect(
+		( select ) => {
+			return select( coreStore ).getEntityRecord(
+				'postType',
+				'wp_template_part',
+				// Ideally this should be an official public API.
+				`${ theme }//${ slug }`
+			);
+		},
+		[ theme, slug ]
+	);
+
+	const linkProps = useLink(
+		{
+			postId: templatePart?.id,
+			postType: templatePart?.type,
+		},
+		{
+			fromTemplateId: params.postId,
+		}
+	);
+
+	if ( ! templatePart ) {
+		return null;
+	}
+
+	return (
+		<BlockControls group="other">
+			<ToolbarButton
+				{ ...linkProps }
+				onClick={ ( event ) => {
+					linkProps.onClick( event );
+				} }
+			>
+				{ __( 'Edit' ) }
+			</ToolbarButton>
+		</BlockControls>
+	);
+}
+
+export const withEditBlockControls = createHigherOrderComponent(
+	( BlockEdit ) => ( props ) => {
+		const { attributes, name } = props;
+		const isDisplayed = name === 'core/template-part' && attributes.slug;
+
+		return (
+			<>
+				<BlockEdit { ...props } />
+				{ isDisplayed && (
+					<EditTemplatePartMenuItem attributes={ attributes } />
+				) }
+			</>
+		);
+	},
+	'withEditBlockControls'
+);
+
+addFilter(
```

Comment:
Interesting choice using the filter here. I don't mind it personally, and like you mention, there's a parallel with the move-to-widget-area. If it weren't for the `useLink` and `useLocation` imports, I'd say it might be good to see if we can somehow get it into the block itself in `block-library`, but maybe not worth the effort of untangling all that since the filter appears to work well 👍

Reply:
> If it weren't for the useLink and useLocation imports...

This is the exact reason I decided to use the filter, plus the "Edit" action is currently Site Editor only.

Reply:
Sounds fair, particularly given that template parts is very site editor centric — so sounds like a reasonable argument for the site editor to have special handling for this block!

Reply:
Thanks for the review, @andrewserong 🙇